File: dat-maven/src/test/kotlin/test/kt/InteropVectorTest.kt
package test.kt

import me.saro.dat.dat.DatManager
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assumptions.assumeTrue
import org.junit.jupiter.api.Test
import java.io.File

/**
 * Parses interop vectors produced by dat-rust (the reference implementation).
 * Vector file path comes from the DAT_INTEROP_VECTORS env var; skipped when absent.
 */
class InteropVectorTest {
    @Test
    fun parseRustVectors() {
        val path = System.getenv("DAT_INTEROP_VECTORS")
        assumeTrue(path != null && File(path).isFile, "DAT_INTEROP_VECTORS not set")

        var certFormat: String? = null
        var manager = DatManager.newInstance()
        var checked = 0

        File(path!!).readLines().filter { it.isNotBlank() }.forEach { line ->
            val (tag, value) = line.split("\t", limit = 2)
            when (tag) {
                "CERT" -> {
                    certFormat = value
                    manager = DatManager.newInstance()
                    manager.imports(value, true)
                }
                else -> {
                    require(certFormat != null) { "vector file must start with CERT" }
                    val payload = manager.parse(value).getOrThrow()
                    when (tag) {
                        "DAT_EMPTY_SECURE" -> {
                            assertEquals("hello", payload.plain)
                            assertEquals("", payload.secure)
                        }
                        "DAT_EMPTY_BOTH" -> {
                            assertEquals("", payload.plain)
                            assertEquals("", payload.secure)
                        }
                        "DAT_NORMAL" -> {
                            assertEquals("hello", payload.plain)
                            assertEquals("world", payload.secure)
                        }
                    }
                    checked++
                }
            }
        }
        assumeTrue(checked > 0, "no dat vectors found")
        println("interop: parsed $checked rust-issued DATs")
    }
}

File: dat-nuget/Saro.Dat.Tests/InteropVectorTest.cs
using Saro.Dat;

namespace Saro.Dat.Tests;

public class InteropVectorTest
{
    [Test]
    public void ParseRustVectors()
    {
        var path = Environment.GetEnvironmentVariable("DAT_INTEROP_VECTORS");
        if (string.IsNullOrEmpty(path) || !File.Exists(path))
        {
            Assert.Ignore("DAT_INTEROP_VECTORS not set");
            return;
        }

        DatManager manager = DatManager.NewInstance();
        var checkedCount = 0;

        foreach (var line in File.ReadAllLines(path).Where(l => !string.IsNullOrWhiteSpace(l)))
        {
            var parts = line.Split('\t', 2);
            var tag = parts[0];
            var value = parts[1];

            if (tag == "CERT")
            {
                manager = DatManager.NewInstance();
                manager.Imports(value, true);
                continue;
            }

            var payload = manager.Parse(value);
            switch (tag)
            {
                case "DAT_EMPTY_SECURE":
                    Assert.That(payload.Plain, Is.EqualTo("hello"));
                    Assert.That(payload.Secure, Is.EqualTo(""));
                    break;
                case "DAT_EMPTY_BOTH":
                    Assert.That(payload.Plain, Is.EqualTo(""));
                    Assert.That(payload.Secure, Is.EqualTo(""));
                    break;
                case "DAT_NORMAL":
                    Assert.That(payload.Plain, Is.EqualTo("hello"));
                    Assert.That(payload.Secure, Is.EqualTo("world"));
                    break;
            }
            checkedCount++;
        }

        Assert.That(checkedCount, Is.GreaterThan(0));
        Console.WriteLine($"interop: parsed {checkedCount} rust-issued DATs");
    }
}
